test/SpagoSpec.hs

```diff
@@ -0,0 +1,158 @@
+import           Test.Hspec          (Spec, around_, beforeAll, describe, it,
+                                      shouldBe)
+import           Turtle              (FilePath, cp, mkdir, mv, readTextFile, rm,
+                                      rmtree, testdir, writeTextFile)
+import           Utils               (checkFixture, runFor, shouldBeFailure,
+                                      shouldBeFailureOutput, shouldBeSuccess,
+                                      shouldBeSuccessOutput, spago, withCwd)
+
+testDir :: FilePath
+testDir = "test/spago-test"
+
+clean :: IO ()
+clean = do
+  whenM (testdir testDir) (rmtree testDir)
+  mkdir testDir
+
+spec :: Spec
+spec = beforeAll clean $ around_ (withCwd testDir) $ do
+
+  describe "spago init" $ do
+
+    it "Spago should have set up a project" $ do
+
+      spago ["init"] >>= shouldBeSuccess
+
+    it "Spago should refuse to overwrite an existing project without -f" $ do
+
+      spago ["init"] >>= shouldBeFailure
+
+    it "Cleaning of config files should succeed" $ do
+
+      rm "spago.dhall" :: IO ()
+      rm "packages.dhall"
+
+    it "Spago should not overwrite files when initing a project" $ do
+
+      writeTextFile "src/Main.purs" "Something"
+      spago ["init"] >>= shouldBeSuccess
+      readTextFile "src/Main.purs" >>= (`shouldBe` "Something")
+      rmtree "src"
+
+    it "Spago should always succeed in doing init with force" $ do
+
+      spago ["init", "-f"] >>= shouldBeSuccess
+
+    it "Spago should import config from psc-package" $ do
+
+      writeTextFile "psc-package.json" "{ \"name\": \"aaa\", \"depends\": [ \"prelude\" ], \"set\": \"foo\", \"source\": \"bar\" }"
+      spago ["init", "-f"] >>= shouldBeSuccess
+      cp "spago.dhall" "spago-psc-success.dhall"
+      checkFixture "spago-psc-success.dhall"
+
+    it "Spago should not import dependencies that are not in the package-set" $ do
+
+      writeTextFile "psc-package.json" "{ \"name\": \"aaa\", \"depends\": [ \"prelude\", \"foo\", \"bar\" ], \"set\": \"foo\", \"source\": \"bar\" }"
+      spago ["init", "-f"] >>= shouldBeSuccess
+      cp "spago.dhall" "spago-psc-failure.dhall"
+      checkFixture "spago-psc-failure.dhall"
+
+
+  describe "spago install" $ do
+
+    it "Subsequent installs should succeed after failed install" $ do
+
+      -- Run `install` once and kill it soon to simulate failure
+      runFor 5000 "spago" ["install", "-j", "3"]
+      -- Sleep for some time, as the above might take time to cleanup old processes
+      threadDelay 1000000
+      spago ["install", "-j", "10"] >>= shouldBeSuccess
+
+    it "Spago should be able to add dependencies" $ do
+
+      spago ["install", "-j10", "simple-json", "foreign"] >>= shouldBeSuccess
+      mv "spago.dhall" "spago-install-success.dhall"
+      checkFixture "spago-install-success.dhall"
+
+    it "Spago should always succeed in doing init with force" $ do
+
+      spago ["init", "-f"] >>= shouldBeSuccess
```

Comment:
I think we can move this init to the next case, since this is basically only "setting up the environment"

Reply:
Done